Test complete run on Anvil

forsyth2 opened the pull request:
Unit test to compare actual and expected images for the all_sets tutorial at https://github.com/E3SM-Project/e3sm_diags/blob/master/examples/tutorials/tutorial_2020_all_sets.py. Due to the large amount of data required to run, this test will be run manually on Anvil rather than as part of the CircleCI tests.

forsyth2 commented on tests/test_complete_run.py:
```diff
@@ -0,0 +1,111 @@
+                path_to_actual_png = os.path.join(actual_images_dir, image)
+                path_to_expected_png = os.path.join(expected_images_dir, image)
+
+                actual_png = Image.open(path_to_actual_png).convert('RGB')
+                expected_png = Image.open(path_to_expected_png).convert('RGB')
+                diff = ImageChops.difference(actual_png, expected_png)
+
+                bbox = diff.getbbox()
+                if not bbox:
+                    # If `diff.getbbox()` is None, then the images are in theory equal
+                    self.assertIsNone(diff.getbbox())
+                else:
+                    # Sometimes, a few pixels will differ, but the two images appear identical.
+                    # https://codereview.stackexchange.com/questions/55902/fastest-way-to-count-non-zero-pixels-using-python-and-pillow
+                    nonzero_pixels = diff.crop(bbox).point(lambda x: 255 if x else 0).convert("L").point(bool).getdata()
+                    num_nonzero_pixels = sum(nonzero_pixels)
+                    print('\npath_to_actual_png={}'.format(path_to_actual_png))
+                    print('path_to_expected_png={}'.format(path_to_expected_png))
+                    print('diff has {} nonzero pixels.'.format(num_nonzero_pixels))
+                    width, height = expected_png.size
+                    num_pixels = width * height
+                    print('total number of pixels={}'.format(num_pixels))
+                    fraction = num_nonzero_pixels/num_pixels
+                    print('num_nonzero_pixels/num_pixels fraction={}'.format(fraction))
+                    # Fraction of mismatched pixels should be less than 0.01%
```
Much of this code is copied from `test_diags.py`. The context it's placed in though is a bit different. I think it's ok to just have this bit of code duplicated here.

forsyth2 commented on tests/test_complete_run.py:
```diff
@@ -0,0 +1,111 @@
+                    if fraction >= 0.0001:
+                        mismatched_images.append(image)
+
+        self.assertEqual(mismatched_images, [])
```
The test will check all images before returning pass/fail, rather than simply failing on the first mismatched image.

forsyth2 reviewed (commented):
@chengzhuzhang This is ready for review.

chengzhuzhang commented on tests/expected_images_complete_run.txt:
```diff
@@ -0,0 +1,475 @@
+./cosp_histogram/MODIS-COSP/MODISCOSP-COSP_HISTOGRAM_MODIS-JJA-global.png
```
I think let's keep the list of files in the Anvil directory. The file size could be too big to include in the repo though. We are still in the process of reducing our repo size...

chengzhuzhang commented on tests/test_complete_run.py:
```diff
@@ -0,0 +1,111 @@
+from acme_diags.parameter.enso_diags_parameter import EnsoDiagsParameter
+from acme_diags.parameter.qbo_parameter import QboParameter
+
+# Due to the large amount of data required to run, this test will be run manually on Anvil
```
Cool, 631 seconds are not too bad.

chengzhuzhang commented on tests/test_complete_run.py:
```diff
@@ -0,0 +1,111 @@
+        runner.sets_to_run = ['lat_lon','zonal_mean_xy', 'zonal_mean_2d', 'polar', 'cosp_histogram', 'meridional_mean_2d', 'enso_diags', 'qbo','area_mean_time_series']
+        runner.run_diags([param, enso_param, qbo_param, ts_param])
+
+        actual_images_dir = './tutorial_2020_all_sets'
```
Use `actual_images_dir = param.results_dir`, so that the test can be run outside the directory?

chengzhuzhang reviewed (commented):
Looks good! Just we need to remind ourselves to upload more expected plots when more sets are being developed.

forsyth2 commented:
@chengzhuzhang Addressed your comments.

I also increased the `num_nonzero_pixels/num_pixels` fraction from 0.0001 (0.01%) to 0.0002 (0.02%).  `JJA_metrics_taylor_diag.png` appears to have sightly more than 0.01% nonzero pixels. I think because there are so many shapes in a small area, any minor movement can cause multiple pixels to change. I inspected the diff visually and I didn't think it was really visible.
